In [6]:
# Load the necessary libraries
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

# Load the dataset
file_path = "D:\DS\Association Rules\Online retail.xlsx"
df = pd.read_excel(file_path)

# Display the first few rows of the dataframe
print(df.head())

# Split the single column into individual items
transactions = df.iloc[:, 0].str.split(',')

# Create a DataFrame where each column represents an item
transactions_df = pd.DataFrame(transactions.tolist())

# Display the transformed dataframe
transactions_df.head()


<>:6: SyntaxWarning: invalid escape sequence '\D'
<>:6: SyntaxWarning: invalid escape sequence '\D'
C:\Users\GAURI DUBEY\AppData\Local\Temp\ipykernel_21560\1107922549.py:6: SyntaxWarning: invalid escape sequence '\D'
  file_path = "D:\DS\Association Rules\Online retail.xlsx"


  shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                            

0          1           2                 3          4     5   \
0         burgers  meatballs        eggs              None       None  None   
1         chutney       None        None              None       None  None   
2          turkey    avocado        None              None       None  None   
3   mineral water       milk  energy bar  whole wheat rice  green tea  None   
4  low fat yogurt       None        None              None       None  None   

     6     7     8     9     10    11    12    13    14    15    16    17  \
0  None  None  None  None  None  None  None  None  None  None  None  None   
1  None  None  None  None  None  None  None  None  None  None  None  None   
2  None  None  None  None  None  None  None  None  None  None  None  None   
3  None  None  None  None  None  None  None  None  None  None  None  None   
4  None  None  None  None  None  None  None  None  None  None  None  None   

     18  
0  None  
1  None  
2  None  
3  None  
4  None

In [7]:
# One-hot encoding of the transactions
basket = transactions_df.stack().reset_index().pivot_table(index='level_0', columns=0, aggfunc='size', fill_value=0)
basket = basket.applymap(lambda x: 1 if x > 0 else 0)

# Display the one-hot encoded dataframe
basket.head()


C:\Users\GAURI DUBEY\AppData\Local\Temp\ipykernel_21560\4075774793.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket = basket.applymap(lambda x: 1 if x > 0 else 0)


0         asparagus  almonds  antioxydant juice  asparagus  avocado  \
level_0                                                               
0                 0        0                  0          0        0   
1                 0        0                  0          0        0   
2                 0        0                  0          0        1   
3                 0        0                  0          0        0   
4                 0        0                  0          0        0   

0        babies food  bacon  barbecue sauce  black tea  blueberries  ...  \
level_0                                                              ...   
0                  0      0               0          0            0  ...   
1                  0      0               0          0            0  ...   
2                  0      0               0          0            0  ...   
3                  0      0               0          0            0  ...   
4                  0      0               0          0            0  ...   

0        turkey  vegetables mix  water spray  white wine  whole weat flour  \
level_0                                                                      
0             0               0            0           0                 0   
1             0               0            0           0                 0   
2             1               0            0           0                 0   
3             0               0            0           0                 0   
4             0               0            0           0                 0   

0        whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
level_0                                                                    
0                        0                 0     0            0         0  
1                        0                 0     0            0         0  
2                        0                 0     0            0         0  
3                        0                 1     0            0         0  
4                        0                 0     0            0         0  

[5 rows x 120 columns]

In [8]:
# Apply the Apriori algorithm with a minimum support threshold
frequent_itemsets = apriori(basket, min_support=0.01, use_colnames=True)

# Generate association rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

# Display the top 10 rules sorted by lift
rules.sort_values('lift', ascending=False).head(10)


C:\Users\GAURI DUBEY\AppData\Local\Programs\Python\Python312\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents                 consequents  \
217               (ground beef)             (herb & pepper)   
216             (herb & pepper)               (ground beef)   
385  (spaghetti, mineral water)               (ground beef)   
388               (ground beef)  (spaghetti, mineral water)   
397  (spaghetti, mineral water)                 (olive oil)   
400                 (olive oil)  (spaghetti, mineral water)   
195                  (tomatoes)         (frozen vegetables)   
194         (frozen vegetables)                  (tomatoes)   
190         (frozen vegetables)                    (shrimp)   
191                    (shrimp)         (frozen vegetables)   

     antecedent support  consequent support   support  confidence      lift  \
217            0.098267            0.049467  0.016000    0.162822  3.291555   
216            0.049467            0.098267  0.016000    0.323450  3.291555   
385            0.059733            0.098267  0.017067    0.285714  2.907540   
388            0.098267            0.059733  0.017067    0.173677  2.907540   
397            0.059733            0.065733  0.010267    0.171875  2.614731   
400            0.065733            0.059733  0.010267    0.156187  2.614731   
195            0.068400            0.095333  0.016133    0.235867  2.474134   
194            0.095333            0.068400  0.016133    0.169231  2.474134   
190            0.095333            0.071333  0.016667    0.174825  2.450820   
191            0.071333            0.095333  0.016667    0.233645  2.450820   

     leverage  conviction  zhangs_metric  
217  0.011139    1.135402       0.772060  
216  0.011139    1.332841       0.732423  
385  0.011197    1.262427       0.697745  
388  0.011197    1.137893       0.727562  
397  0.006340    1.128171       0.656783  
400  0.006340    1.114306       0.661001  
195  0.009613    1.183913       0.639564  
194  0.009613    1.121370       0.658605  
190  0.009866    1.125418       0.654355  
191  0.009866    1.180480       0.637444

In [10]:
# Analyzing the top 10 rules
top_rules = rules.sort_values('lift', ascending=False).head(10)

# Display the top rules
print(top_rules)


                    antecedents                 consequents  \
217               (ground beef)             (herb & pepper)   
216             (herb & pepper)               (ground beef)   
385  (spaghetti, mineral water)               (ground beef)   
388               (ground beef)  (spaghetti, mineral water)   
397  (spaghetti, mineral water)                 (olive oil)   
400                 (olive oil)  (spaghetti, mineral water)   
195                  (tomatoes)         (frozen vegetables)   
194         (frozen vegetables)                  (tomatoes)   
190         (frozen vegetables)                    (shrimp)   
191                    (shrimp)         (frozen vegetables)   

     antecedent support  consequent support   support  confidence      lift  \
217            0.098267            0.049467  0.016000    0.162822  3.291555   
216            0.049467            0.098267  0.016000    0.323450  3.291555   
385            0.059733            0.098267  0.017067    0.285714  2.

1. The association rule mining on the dataset reveals several strong product relationships and purchasing patterns Notably "Herb & pepper" and "ground beef" have a high lift of 3.29 indicating that they are frequently bought together Similarly , "ground beef" often co-occurs with "mineral water and spaghetti" and "frozen vegetables' show strong associations with "tomatoes" and "shrimp"
2. These insights suggest valuable cross=selling and bundling pooprtunities for instance promoting "herb & pepper" with "ground beef" or offering bundle deals for "mineral water , spaghetti and ground beef" could boost sales  additionaly these findings can inform inventory management and product placement strategies ensuring that frequently associated items are stocked together and easily accessible to customers
3. Overall understanding these purchasing behaviours can help retailers enchance customer satisfation optimize marketing efforts and increase revenue through strategic product placements and promotions 

**Interview Questions:**
1.	What is lift and why is it important in Association rules?
Answer : Lift measures the strength of an association rule compared to the expected frequency of the items occurring together by chance It is calculated as 
        Lift(A --> B) = Support(A)×Support(B)/Support(A∪B)
A lift greater than 1 indicates a positive correlation, meaning the items are likely to be purchases together more frequency than by random chance Lift is important because it help identify strong meaningful associations between items 

2. What is support and Confidence. How do you calculate them?
Answer :         a.Support is the proportion of transactions that contain a specific itemset It is calculated as
 **Support(A) = Number of transactions containing A / Total number of transactions**
                 b.Confidence measures like likelihood that an item B is purchases when otem A is purchased It is calculated as
   **Confidence(A-->B) = Support(A∪B)/Support(A)**

3.	What are some limitations or challenges of Association rules mining?
Answer :
            a. Scalabilty : Handling large datasets can be computationally intensive
            b. Rare item problem : Rare but interesting itemsets might be missed due to low support threshold
            c.Interpretability : A large number of generated rules can be overwhelming and difficult to interpret
            d.Redundancy: Many rules might be redundant containing similar information 
